libraries/core_libs/consensus/src/pbft/round.cpp

```diff
@@ -0,0 +1,203 @@
+#include "pbft/round.hpp"
+
+#include "pbft/pbft_manager.hpp"
+#include "pbft/period_data.hpp"
+#include "pbft/step/certify.hpp"
+#include "pbft/step/filter.hpp"
+#include "pbft/step/finish.hpp"
+#include "pbft/step/polling.hpp"
+#include "pbft/step/propose.hpp"
+#include "vote_manager/vote_manager.hpp"
+
+namespace taraxa {
+
+Round::Round(uint64_t id, std::shared_ptr<NodeFace> node) : RoundFace(id, std::move(node)) {
+  const auto& node_addr = getNode()->node_addr_;
+  LOG_OBJECTS_CREATE("ROUND_" + std::to_string(getId()));
+  previous_round_next_voted_value_ = node_->next_votes_manager_->getVotedValue();
+  previous_round_next_voted_null_block_hash_ = node_->next_votes_manager_->haveEnoughVotesForNullBlockHash();
+}
+
+Round::~Round() { step_.reset(); }
+
+std::shared_ptr<Round> Round::make(uint64_t id, std::optional<uint64_t> step, std::shared_ptr<NodeFace> node) {
+  auto round = std::make_shared<Round>(id, node);
+  round->start(step);
+
+  return round;
+}
+
+void Round::run() {
+  auto time_from_round_start = std::chrono::system_clock::now() - start_time_;
+  time_from_start_ms_ = std::chrono::duration_cast<std::chrono::milliseconds>(time_from_round_start).count();
+  step_->run();
+  if (!step_->finished()) {
+    sleepUntil(std::chrono::milliseconds(time_from_start_ms_ + getLambda() / 4));
+  } else {
+    nextStep_();
+  }
+}
+
+void Round::sleepUntil(const time_point& time) {
+  std::unique_lock<std::mutex> lock(stop_mtx_);
+  stop_cv_.wait_until(lock, time);
+}
+
+void Round::sleepUntil(const std::chrono::milliseconds& time) { sleepUntil(start_time_ + time); }
+
+StepType Round::getStepType() const { return step_->getType(); }
+
+void Round::finish() {
+  finished_ = true;
+  {
+    std::unique_lock<std::mutex> lock(stop_mtx_);
```

Comment:
maybe I am wrong but do we need to lock it here?

Reply:
Yeah, looks like we don't need this. Just did it as it was in pbft_manager without any thoughts :)